In [32]:
cnt =0
for n in range(1000, 10000):
    num =[]
    s = ''
    while n >0:
        num.append(n% 10)
        n //=10
    
    num.sort()
    s = str(num[0] + num[3]) + str(num[1]*num[2])
    if int(s) == 1128:
        cnt += 1

cnt
    

54

In [37]:
cnt = 0

for i in range(300, 401):
    num = []
    while i > 0:
        num.append(i% 10)
        i //=10
    num = sorted(num, reverse=True)
    maxi = str(num[0]) + str(num[1])
    if num[2] != 0:
        mini = str(num[2]) + str(num[1])
    else:
        mini = str(num[1]) + str(num[0])
    if int(maxi)-int(mini) == 20:
        cnt +=1

cnt

12

In [3]:
def f(n):
    s = ''
    while n >0:
        s = str(n % 8) + s
        n //=8
    return s

def g(n):
    c = 0
    while n >0:
        c += n % 10
        n //=10
    return c

r = []

for i in range(483, 1000):
    m = f(i)
    if g(int(m)) % 2 ==0:
        r.append(int(m + f(g(int(m))), 8))
    else:
        r.append(int(f(g(int(m))) + m, 8))

r.sort()
r

[4106,
 4124,
 4142,
 4162,
 4180,
 4198,
 4236,
 4254,
 4292,
 4310,
 4366,
 4422,
 4608,
 4610,
 4628,
 4646,
 4684,
 4702,
 4740,
 4758,
 4814,
 4870,
 5132,
 5150,
 5188,
 5206,
 5262,
 5318,
 5636,
 5654,
 5710,
 5766,
 6158,
 6214,
 6662,
 7145,
 7152,
 8164,
 8171,
 8178,
 8185,
 9190,
 9197,
 9204,
 9211,
 10223,
 10230,
 10237,
 11263,
 12802,
 12809,
 12816,
 12865,
 12872,
 12928,
 20996,
 21003,
 21010,
 21017,
 21024,
 21059,
 21066,
 21073,
 21080,
 21122,
 21129,
 21136,
 21185,
 21192,
 21248,
 29190,
 29197,
 29204,
 29211,
 29218,
 29225,
 29232,
 29253,
 29260,
 29267,
 29274,
 29281,
 29288,
 29316,
 29323,
 29330,
 29337,
 29344,
 29379,
 29386,
 29393,
 29400,
 29442,
 29449,
 29456,
 29505,
 29512,
 29568,
 30926,
 31056,
 31186,
 31244,
 31374,
 31504,
 31634,
 31822,
 31952,
 32082,
 32212,
 32270,
 32400,
 32530,
 32660,
 33224,
 33672,
 34120,
 34250,
 34568,
 34698,
 35016,
 35146,
 35276,
 35464,
 35594,
 35724,
 35912,
 36042,
 36172,
 36302,
 36360,
 3649

In [3]:
def tr(n):
    s = ''
    while n >0:
        s = str(n % 3) + s
        n //=3
    return s

for N in range(1, 60000):

    n = tr(N)
    s = ''
    if N % 3 ==0:
        for i in range(len(n)):
            s += n[i]
            s += n[i]
        n = s
    else:
        n = n.replace('0', 'q').replace('1', 'w').replace('2', 'e')
        n = n.replace('q', '1').replace('w', '2').replace('e', '0')
        for i in range(len(n)):
            s += n[i]
            s += n[i]
        n = s
    if int(n, 3) > 120:
        print(N)
        break

9


На вход алгоритма подаётся натуральное число N. Алгоритм строит по нему новое число R следующим образом.

1.  Строится троичная запись числа N.

2.  Далее эта запись обрабатывается по следующему правилу:

а)  если число N делится на 3, то к этой записи справа дописываются две последние троичные цифры полученной записи;

б)  если число N на 3 не делится, то вычисляется сумма цифр полученной троичной записи, эта сумма умножается на 3, переводится в троичную систему счисления и дописывается в конец числа.

Полученная таким образом запись является троичной записью искомого числа R.

3.  Результат переводится в десятичную систему и выводится на экран.

Например, для исходного числа 8 = 223 результатом является число 221103   =  228, а для исходного числа 9  =  1003 результатом является число 100003  =  81.

Укажите число R, ближайшее к 910, которое может быть получено с помощью описанного алгоритма. В ответе запишите это число в десятичной системе счисления.

In [10]:
def f(n):
    s= ''
    while n >0:
        s = str(n%3) + s
        n //= 3
    return s

for N in range(1, 10000):
    n = f(N)
    if N % 3 == 0:
        n = n + n[-2:]
    else:
        a= int(n)
        ss = 0
        while a > 0:
            ss += a % 10
            a//=10
        n= n + f(3*ss)
    if 860<int(n, 3) < 950:
        print(int(n, 3))


876
930
870
891
921


На вход алгоритма подаётся натуральное число N. Алгоритм строит по нему новое число R следующим образом.

1.  Строится троичная запись числа N.

2.  Далее эта запись обрабатывается по следующему правилу:

а)  если сумма цифр троичной записи числа N делится на 3, то в этой записи два левых разряда заменяются на «112»;

б)  если сумма цифр троичной записи числа N на 3 не делится, то эта сумма переводится в троичную систему счисления и дописывается в конец числа.

Полученная таким образом запись является троичной записью искомого числа R.

3.  Результат переводится в десятичную систему и выводится на экран.

Например, для исходного числа 11  =  1023 результатом является число 11223  =  44, а для исходного числа 12  =  1103 результатом является число 11023  =  38.

Укажите минимальное чётное число R, большее 702, которое может быть получено с помощью описанного алгоритма. В ответе запишите это число в десятичной системе счисления.

In [16]:
def f(n):
    s= ''
    while n >0:
        s = str(n%3) + s
        n //= 3
    return s
d = []
for N in range(1, 1000000):

    n = f(N)
    a= int(n)
    ss = 0
    while a > 0:
        ss += a % 10
        a//=10
    if ss % 3 == 0:
        n = '112' + n[2:]
    else:
        n = n + f(ss)
    if int(n, 3) % 2 ==0:
        if int(n, 3) > 702:
            d.append(int(n, 3))

min(d)
    


718

1.  Строится троичная запись числа N.

2.  В полученной записи все нули заменяются на двойки, все двойки  — на нули. Из полученного числа удаляются ведущие нули.

3.  Результат переводится в десятичную систему счисления.

4.  Результатом работы алгоритма становится модуль разности исходного числа N и числа, полученного на предыдущем шаге.

 

Пример. Дано число N  =  35. Алгоритм работает следующим образом.

1.  Строим троичную запись числа N: 3510  =  10223.

2.  Заменяем цифры и удаляем ведущие нули: 1022 → 1200.

3.  Переводим в десятичную систему: 12003  =  4510.

4.  Вычисляем модуль разности: |35 − 45|  =  10.

Результат работы алгоритма R  =  10.

При каком наименьшем N в результате работы алгоритма получится R  =  1 864 246.

In [18]:
def f(n):
    s= ''
    while n >0:
        s = str(n%3) + s
        n //= 3
    return s

for N in range(1, 10000000000):
    n = f(N).replace('2', 'q').replace('0', 'e')
    n = n.replace('q', '0').replace('e', '2')
    r = abs(N - int(n, 3))
    if r == 1864246:
        print(N)
        break


3323607


Автомат получает на вход четырёхзначное десятичное число, в котором все цифры нечётные. По этому числу строится новое число по следующим правилам.

1.  Складываются первая и вторая, а также третья и четвёртая цифры.

2.  Полученные два числа записываются друг за другом в порядке неубывания (без разделителей).

Пример. Исходное число: 7511. Суммы: 7 + 5  =  12; 1 + 1  =  2. Результат: 212. Сколько существует чисел, в результате обработки которых автомат выдаст число 414.

In [25]:
cnt = 0
for i in range(1000, 10000):
    s= str(i)
    if int(s[0]) % 2 == 1:
        if int(s[1]) % 2 == 1: 
            if int(s[2]) % 2 == 1: 
                if int(s[3]) % 2 == 1: 
                    q = int(s[0]) + int(s[1])
                    w = int(s[2]) + int(s[3])
                    r = str(sorted([q,w])[0]) + str(sorted([q,w], reverse=True)[0])

                    if r == '414':
                        cnt +=1
cnt

12

In [6]:

def f(n):
    s= ''
    while n >0:
        s = str(n%3) + s
        n //= 3
    return s

for N in range(1, 1000):
    n = f(N)
    if n[0] == n[-1]:
        n+=n[0]
    else:
        n += str(max(int(n[0]), int(n[-1])))
    r= int(n, 3)
    if r >49:
        print(N)
        break
        

17


In [38]:
d= []
for N in range(42, 1000):
    n = bin(N)[2:]
    if N % 2 !=0:
        n = '10' + n[:-2] + '01'
    else:
        n = '11' + n[2:] + '1'
    d.append(int(n, 2))

min(d)

97

In [35]:
s = '1234'
s[:-2]

'12'

In [6]:
ns = []
for N in range(1, 100000):
    n = bin(N)[2:]
    if N % 2 ==0:
        n = '10' + n
    else:
        n = '1' + n + '01'
    r=int(n, 2)
    if r <30:
        ns.append(N)
    
max(ns)

6

In [14]:
def f(n):
    s = ''
    while n >0:
        s = str(n % 7) + s
        n//=7
    return s
ns = []
for N in range(1, 10000000):
    n = f(N)
    if n.count('2') % 2 ==0:
        n += '555'
    else:
        n = '1' + n
    r = int(n, 7)
    if r < 3799:
        ns.append(N)
max(ns)


1395

In [24]:
rs =[]

for N in range(1, 1000000):
    n = bin(N)[2:]
    if N % 3 ==0:
        n = n + n[-2:]
    else:
        n = n +str(bin((N % 3)*3)[2:])
    r = int(n, 2)
    if r >= 195:
        rs.append(r)
min(rs)

199

In [ ]:
def f(n):
    s= ''
    while n >0:
        s = str(n%3) + s
        n //= 3
    return s
rs = []
for N in range(1, 100000):
    n = f(N)
    if N % 3 == 0:
        n += n[-2:]
    else:
        n += f((N% 3) * 5)
    r = int(n, 3)
    if r > 133:
        rs.append(r)
min(rs)

141

In [2]:
nums =[]
for N in range(1000, 10000):
    n = str(N)
    a = (int(n[0]) * int(n[1]))
    b = (int(n[1]) * int(n[2]))
    c = (int(n[2]) * int(n[3]))
    parts = sorted([a, b, c])
    s = "".join(str(x) for x in parts)
    s = s.lstrip('0')
    if len(s) == 3:
        nums.append(int(s))
        

min(nums)

111

In [16]:
for N in range(1000, 10000):

    n = str(N)
    a= int(n[0]) + int(n[1])
    b = int(n[2]) + int(n[3])
    num = sorted([a, b])
    s = ''.join(str(x) for x in num)
    num = [int(x) for x in s]
    d = str(max(num) - min(num))
    s += d
    if s == '17176':
        print(N)
        break

8989


In [17]:

for N in range(100, 1000):
    n= str(N)
    a = str(int(n[0])**2 + int(n[1])**2)
    b = str(int(n[1])**2 + int(n[2])**2)
    num = sorted([a,b], reverse=True)
    s = ''.join(str(x) for x in num)
    if int(s) == 9010:
        print(N)
        break

139


In [39]:
for i in range(131, 256):
    n = bin(i)[2:].zfill(8)
    s = n[:2] + n[6:]
    if int(s, 2) == 10:
        print(i)
        break

134
